language/move-vm/runtime/src/loader.rs

```diff
@@ -624,30 +629,62 @@ impl Loader {
         Self::check_natives(&module)?;
 
         let deps = module.immediate_module_dependencies();
-        let loaded_deps = if verify_no_missing_modules {
+        let loaded_imm_deps = if verify_no_missing_modules {
             self.load_dependencies_verify_no_missing_dependencies(deps, data_store, log_context)?
         } else {
             self.load_dependencies_expect_no_missing_dependencies(deps, data_store, log_context)?
         };
 
-        self.verify_module_dependencies(module, loaded_deps)
+        self.verify_module_dependencies(
+            module,
+            loaded_imm_deps,
+            self.get_all_module_dependencies(module),
+        )
     }
 
     fn verify_module_dependencies(
         &self,
         module: &CompiledModule,
-        dependencies: Vec<Arc<Module>>,
+        imm_dependencies: Vec<Arc<Module>>,
+        all_dependencies: HashMap<ModuleId, Arc<Module>>,
     ) -> VMResult<()> {
-        let mut deps = vec![];
-        for dep in &dependencies {
-            deps.push(dep.module());
+        let imm_deps: Vec<_> = imm_dependencies
+            .iter()
+            .map(|module| module.module())
+            .collect();
+        DependencyChecker::verify_module(module, imm_deps)?;
+
+        let all_deps = all_dependencies
+            .iter()
+            .map(|(module_id, module)| (module_id.clone(), module.module()))
+            .collect();
+        CyclicModuleDependencyChecker::verify_module(module, all_deps)
+    }
+
+    fn get_all_module_dependencies(
+        &self,
+        module: &CompiledModule,
+    ) -> HashMap<ModuleId, Arc<Module>> {
```

Comment:
Rust nit: I generally default to `BTreeMap` unless I have a very compelling argument for using `HashMap`. The reason: iteration over `HashMap`'s is not deterministic, but iteration over `BTreeMap`s is deterministic because keys implement `Ord`. 

That nondeterminism often bites in unexpected ways. For example: we might detect different missing dependencies in different runs of the VM on the same input, and the name of the missing dependency might make its way into an error message.

Reply:
Will make an update on this!